rig/bitfield.py

```diff
@@ -187,41 +203,38 @@ def __call__(self, **field_values):
         ValueError
             If any field has already been assigned a value or the value is too
             large for the field.
-        AttributeError
-            If a field is specified which is not present.
+        UnavailableFieldError
+            If a field is specified which does not exist or is not available.
         """
-        # Ensure fields exist
-        for identifier in field_values.keys():
-            if identifier not in self.fields:
-                raise ValueError("Field '{}' not defined.".format(identifier))
-
         # Make sure no values are changed
         for identifier, value in self.field_values.items():
             if identifier in field_values:
                 raise ValueError(
                     "Field '{}' already has value.".format(identifier))
 
+        # Work out what the new set of defined values looks like
         field_values.update(self.field_values)
 
-        # Ensure no fields are specified which are not enabled
-        for identifier in field_values:
-            self._assert_field_available(identifier, field_values)
-
-        # Ensure values are within range
         for identifier, value in field_values.items():
-            field_length = self.fields[identifier].length
+            # All fields specified must exist
+            field = self.fields.get_field(identifier, field_values)
+
+            # All field values must be within range
             if value < 0:
                 raise ValueError("Fields must be positive.")
-            elif field_length is not None and value >= (1 << field_length):
+            elif field.length is not None and value >= (1 << field.length):
```

Comment:
Unfortunately it makes behaviour when (unsuported) lengths of 0 are used very odd. At present I think lengths of zero will just work.

Further, it makes the rule of "if you don't have a fixed length, you don't get an exception until you assign lengths" more complicated. I think cases where a bug gets you a value < 1<<32 (typical case) should be rare enough not to warrant extra help.

What do you think?